pkg/provider/infisical/provider_test.go

```diff
@@ -0,0 +1,236 @@
+				apiScope:  &apiScope,
+			}
+			var property string
+			if tc.Name == "Get_property_key" {
+				property = "key"
+			}
+
+			output, err := p.GetSecret(context.Background(), esv1beta1.ExternalSecretDataRemoteRef{
+				Key:      "key",
+				Property: property,
+			})
+
+			if tc.Error == nil {
+				assert.NoError(t, err)
+				assert.Equal(t, tc.Output, output)
+			} else {
+				assert.ErrorAs(t, err, &tc.Error)
+			}
+		})
+	}
+}
+
+func TestGetSecretMap(t *testing.T) {
+	testCases := []TestCases{
+		{
+			Name: "Get_valid_key_map",
+			MockClient: &fake.MockInfisicalClient{
+				MockedGetSecretByKeyV3: func(data api.GetSecretByKeyV3Request) (string, error) {
+					return `{"key":"value"}`, nil
+				},
+			},
+			Error: nil,
+			Output: map[string][]byte{
+				"key": []byte("value"),
+			},
+		},
+		{
+			Name: "Get_invalid_map",
+			MockClient: &fake.MockInfisicalClient{
+				MockedGetSecretByKeyV3: func(data api.GetSecretByKeyV3Request) (string, error) {
+					return ``, nil
+				},
+			},
+			Error:  errors.New("unexpected end of JSON input"),
+			Output: nil,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.Name, func(t *testing.T) {
+			p := &Provider{
+				apiClient: tc.MockClient,
+				apiScope:  &apiScope,
+			}
+			output, err := p.GetSecretMap(context.Background(), esv1beta1.ExternalSecretDataRemoteRef{
+				Key: "key",
+			})
+			if tc.Error == nil {
+				assert.NoError(t, err)
+				assert.Equal(t, tc.Output, output)
+			} else {
+				assert.ErrorAs(t, err, &tc.Error)
+			}
+		})
+	}
+}
+
+func makeSecretStore(projectSlug, environment, secretPath string, fn ...storeModifier) *esv1beta1.SecretStore {
+	store := &esv1beta1.SecretStore{
+		Spec: esv1beta1.SecretStoreSpec{
+			Provider: &esv1beta1.SecretStoreProvider{
+				Infisical: &esv1beta1.InfisicalProvider{
+					Auth: esv1beta1.InfisicalAuth{
+						UniversalAuthCredentials: &esv1beta1.UniversalAuthCredentials{},
+					},
+					SecretsScope: esv1beta1.MachineIdentityScopeInWorkspace{
+						SecretsPath:     secretPath,
+						EnvironmentSlug: environment,
+						ProjectSlug:     projectSlug,
+					},
+				},
+			},
+		},
+	}
+	for _, f := range fn {
+		store = f(store)
+	}
+	return store
+}
+
+func withClientID(name, key string, namespace *string) storeModifier {
+	return func(store *esv1beta1.SecretStore) *esv1beta1.SecretStore {
+		store.Spec.Provider.Infisical.Auth.UniversalAuthCredentials.ClientID = esv1meta.SecretKeySelector{
+			Name:      name,
+			Key:       key,
+			Namespace: namespace,
+		}
+		return store
+	}
+}
+
+func withClientSecret(name, key string, namespace *string) storeModifier {
+	return func(store *esv1beta1.SecretStore) *esv1beta1.SecretStore {
+		store.Spec.Provider.Infisical.Auth.UniversalAuthCredentials.ClientSecret = esv1meta.SecretKeySelector{
+			Name:      name,
+			Key:       key,
+			Namespace: namespace,
+		}
+		return store
+	}
+}
+
+type ValidateStoreTestCase struct {
+	store *esv1beta1.SecretStore
+	err   error
```

Comment:
A nicer way to do error checking is simply pass in a function like this:

```go
assertError func(t *testing.T, err)
```

and then in the test:

```go
		{
			store: makeSecretStore(apiScope.ProjectSlug, apiScope.EnvironmentSlug, apiScope.SecretPath, withClientSecret(authType, randomID, nil)),
			err:   func(t *testing.T, err) { require.NoError(t, err) },
		},
```

And then in the test :

```go
		_, err := p.ValidateStore(tc.store)
                tc.assertError(t, err)
```

This way, you avoided the convoluted checking based on what the error was, and made sure the test has control over what it wanted. :)

Reply:
Done